stack.py

```diff
@@ -168,9 +160,90 @@ def parse_program(code):
         yield Instr(op, args, prefix)
 
 
+def is_well_formed(parts):
+    r"""
+    Take an instruction split into a list of parts and check if it is well formed
+    Does not typecheck the instructions, only that they make syntaxtical sense
+    It also doesn't care if a particular combination of prefixes and op aren't valid,
+    so it will pass 'add @label' even though that doesn't actually work
+
+    >> is_well_formed(['@label'])
+    True
+
+    >> is_well_formed(['@label', 'add'])
+    False
+
+    >> is_well_formed(['jump', 'quiet'])
+    False
+    """
+    # Check if valid naked label
+    if is_label(parts[0]):
+        if len(parts) == 1:
+            return True
+        else:
+            return False
+
+    # Since it isn't a naked label, the first part must be a prefix or an op
+    if not (parts[0] in prefixes or is_op(parts[0])):
+        return False
+
+    # No op or more then one op is invalid.
+    if sum(is_op(part) for part in parts) != 1:
+        return False
+
+    # Find the positions of the parts.
+    for i, part in enumerate(parts):
```

Comment:
A simpler algorithm for this. Keep track of:
- A list of prefix parts
- Whether the op has been found
- Whether the label has been found
As you go:
- If you find a prefix part, check if it's in the list already. If it is, `return False`, otherwise add it to the list. If you've already found either the label or the op, this is an error.
- If it's an op, keep track of the fact that you found the op. If you've already found either the label or an op, this is an error.
- If it's a label, keep track of the fact that you found the label. If you've already found the label, this is an error.